### ISSUE TITLE
Combination of remove and filter causes stack overflow

### ISSUE THREAD

**Issue by matentzn** [ontodev/robot#527]
The test is here (ROBOT 1.4.1)

```
git clone https://github.com/monarch-ebi-dev/robot_test.git
cd robot_test
make fail-remove-filter.owl
```
The command run is:

```
robot remove --input input-fail-remove-filter.owl --axioms equivalent \
		filter --term-file simple_seed.txt --trim true --signature true --output fail-remove-filter.owl
```

The error message is:

```
Exception in thread "main" java.lang.StackOverflowError
	at com.carrotsearch.hppcrt.maps.ObjectObjectHashMap.hashKey(ObjectObjectHashMap.java:136)
	at com.carrotsearch.hppcrt.maps.ObjectObjectHashMap.get(ObjectObjectHashMap.java:779)
	at uk.ac.manchester.cs.owl.owlapi.MapPointer.get(MapPointer.java:382)
	at uk.ac.manchester.cs.owl.owlapi.MapPointer.getValues(MapPointer.java:191)
	at uk.ac.manchester.cs.owl.owlapi.OWLImmutableOntologyImpl.getAxioms(OWLImmutableOntologyImpl.java:1325)
	at uk.ac.manchester.cs.owl.owlapi.OWLAxiomIndexImpl.getSubClassAxiomsForSubClass(OWLAxiomIndexImpl.java:136)
	at uk.ac.manchester.cs.owl.owlapi.concurrent.ConcurrentOWLOntologyImpl.getSubClassAxiomsForSubClass(ConcurrentOWLOntologyImpl.java:1768)
	at org.semanticweb.owlapi.search.EntitySearcher.getSuperClasses(EntitySearcher.java:805)
	at org.obolibrary.robot.RelatedObjectsHelper.getSuperClasses(RelatedObjectsHelper.java:1014)
	at org.obolibrary.robot.RelatedObjectsHelper.spanGapsHelper(RelatedObjectsHelper.java:1407)
```

What is weird that only the combination of these two commands causes the problem; neither in isolation does!

**Comment by matentzn** [comment]
Maybe related (this is supposed to remove everything except for some class in term.txt, its descendants and its labels):

```
robot remove -i scAtlas.owl -T term.txt --term rdfs:label --select “self descendants” --select complement -o test.owl
```

Results in:

```
Exception in thread “main” java.lang.StackOverflowError
    at java.util.AbstractCollection.toArray(AbstractCollection.java:137)
    at java.util.ArrayList.<init>(ArrayList.java:178)
    at uk.ac.manchester.cs.owl.owlapi.MapPointer.get(MapPointer.java:386)
    at uk.ac.manchester.cs.owl.owlapi.MapPointer.getValues(MapPointer.java:191)
    at uk.ac.manchester.cs.owl.owlapi.OWLImmutableOntologyImpl.getAxioms(OWLImmutableOntologyImpl.java:1325)
    at uk.ac.manchester.cs.owl.owlapi.OWLAxiomIndexImpl.getSubClassAxiomsForSuperClass(OWLAxiomIndexImpl.java:142)
    at uk.ac.manchester.cs.owl.owlapi.concurrent.ConcurrentOWLOntologyImpl.getSubClassAxiomsForSuperClass(ConcurrentOWLOntologyImpl.java:1779)
    at org.semanticweb.owlapi.search.EntitySearcher.getSubClasses(EntitySearcher.java:839)
    at org.obolibrary.robot.RelatedObjectsHelper.selectClassDescendants(RelatedObjectsHelper.java:1218)
    at org.obolibrary.robot.RelatedObjectsHelper.selectClassDescendants(RelatedObjectsHelper.java:1223)
```

**Comment by matentzn** [comment]
Could it be that EntitySearcher does not deal well with cycles? Maybe its better to use OWLStructuralReasoner instead.. Not sure.

**Comment by rctauber** [comment]
Yes, it seems that the issue is with the cycles after removing all the equivalent axioms.

I have looked into using `OWLStructuralReasoner`, but there are a few problems. Unlike `EntitySearcher`, it only returns named classes, not anonymous superclasses. I worry that some important axioms may be lost if we switch to this. Also, using a reasoner with your test ontology instead of `EntitySearcher` still causes a stack overflow on `reasoner.getSuperClasses(...)`.

I've been playing around with this issue all weekend and the issue the only solution I can think of right now is to throw a more-detailed exception telling the user to use `--preserve-structure false` if there are cycles. Running this command succeeds and there is still *some* hierarchy, but gaps will not be filled in:

```
robot remove --input input-fail-remove-filter.owl --axioms equivalent \
    filter --term-file simple_seed.txt --select "self annotations" --trim true \
    --signature true --preserve-structure false --output fail-remove-filter.owl
```

**Comment by matentzn** [comment]
Can you try to describe what exactly is going wrong here? Maybe I can think of something... :P

**Comment by rctauber** [comment]
Getting either the equivalent classes or the superclasses for 'chemical entity' and 'chemical' (which are subclasses of each other) causes a stack overflow due to the cycle. I had some trouble identifying this issue but that seems to be the main cause. It causes a stack overflow both with the `EntitySearcher` and the `OWLReasoner`.

This is happening during the step where gaps between subclasses are filled in (`spanGaps`) after filtering/removing. This is to preserve the structure of the hierarchy, even if you got rid of a class between two, e.g.:
```
- parent
    - child
        - child 2
```
 Becomes this if you remove `child`:
```
- parent
    - child 2
```

For big and complicated ontologies, `spanGaps` can be pretty inefficient because it has to look at the ancestors of each class and find which ones are in the output, and then create `subClassOf` axioms for the new parent-child relationships. `spanGaps` also runs into issues with these circular definitions, since it will get stuck in a loop. 

I thought I had made the appropriate fixes by removing any instances where a class is a subclass of itself, or a class is both equivalent and subclass. But, it looks like since 'chemical entity' and 'chemical' are subclasses of each other, but the equivalent statement is no longer there, ROBOT doesn't know to remove it and it gets stuck in a loop. But, the main problem is that just *getting* the superclasses is causing a stack overflow, which originates in OWLAPI.

Here's the related code:
https://github.com/ontodev/robot/blob/master/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java#L1012

I've tried putting in try/catch blocks in a test branch, but I'm still running into problems. Also, try/catch just seems like a bandaid for a bigger issue.

Hopefully that provided a good overview, but let me know if you have any other questions.

**Comment by matentzn** [comment]
Ok, I think from all this it is clear we need our own hierarchy browser. I will make a new ticket for that.

**Comment by jamesaoverton** [comment]
I don't see how this is a ROBOT bug. Either

- OWLAPI should not overflow on cycles, or
- ontologies should not contain subClassOf cycles

We tried to work around the problem in ROBOT, but if we can't get the superclasses then we're stuck.

**Comment by matentzn** [comment]
Fair enough; I do not think we can assume ontologies to not have these cycles; they will, and I would like to use ROBOT on them. I will try to see what Ignazio says about this, and whether this is something that can be fixed on OWLAPI end.

**Comment by matentzn** [comment]
@rctauber I was just about to make a ticket at the OWL API, but I think I missed something; (1) is it spanGaps that overflows or EntitySearcher? I think its spanGaps! (2) what SHOULD be the behaviour of EntitySearcher.getSuperclasses() if we have:

A sub B
B sub A

A sub B
B sub C
C sub A

To avoid cycles? Can this even be done at all?

@jamesaoverton Are you suggesting that ROBOT should simply not work with ontologies that contain 

```
A sub B
B sub A
```

Here is a simple example:

```
git clone https://github.com/monarch-ebi-dev/robot_test.git
cd robot_test
make fail-abc-eq.owl
```

This query: 

```
	robot relax -i abc.owl remove --axioms equivalent remove --term http://www.purl.obolibrary.org/obo/ZP_0099004 -o $@.owl
```

Is run against an ontology that is cycle free:

O1:

```
A=B=C
```

And Robots spanGaps stack-overflows. 

Or in other words: whenever we combine relax with remove equivalents we have a risk of hitting this spot. It seems like a very usual thing to do..

In any case, some people convinced me in private conversion that ROBOT follows a structural approach, and perhaps using a reasoner is not the answer here. But handling the simple case above seems to be important to me, but correct me if I am wrong.

**Comment by jamesaoverton** [comment]
I think you can always find some way to avoid cycles. The problem is that there's usually more than one way to do it, and some ways are resource intensive.

It makes sense to assert `A=B=C`, and if ROBOT fails on that then it's a bug.

I can't think of a good reason for asserting `A sub B, B sub A`. I'd love to see a convincing example / use case. If I saw this in the wild, my first thought would be that the editor made a mistake. If it's not a mistake, then it's confusing, and the editor should just assert `A=B`. I'm willing to be convinced otherwise.

My main concern is resources. @rctauber and I are way, way, way over budget on ROBOT this year and we need to scale back. `A=B=C` is a legitimate case. `A sub B, B sub A` seems like an edge case to me. Some of the solutions proposed above sound expensive. So I'm trying to figure out the cost/benefit.

In any case, we need to figure out / decide where the problem really is: OWLAPI, ROBOT, or the ontology. I honestly don't know. But I think we're making some progress here, and I appreciate the time you're putting in to this.

**Comment by matentzn** [comment]
Cool, lets take this slowly :). I think the resource argument is very serious, but I will leave this to @cmungall to figure out. It was my understanding that I should very seriously try to replace all use of owltools with ROBOT everywhere; maybe I am too overzealous with that and @cmungall should reign me in again. So just for the record: The *only place* I currently can't use ROBOT is to generate these [simple and basic release files](https://github.com/INCATools/ontology-development-kit/pull/250/files) as we have specified them (I use some strange mix of owltools and ROBOT to achieve the goal for ontologies such as CL and FBCV). For everything else, it works perfectly fine! 

Lets gather some key issues here and focus on those, to save resources:

`F1: If you relax an ontology that contains named equivalence, you will necessarily create a subclass of cycle. `

So relaxing `A=B` will create the `A sub B, B sub A` ontology. So regardless of whether the A sub B, B sub A situation is asserted and that makes sense (I personally agree with you that asserting it is not useful and will happen rarely; A sub B, B sub A is just a very wordy way to say the exact same A=B statement), relax will just do this, whether you want it or not..  Does this mean, that maybe, relax should not relax equivalences between named classes (adding another parameter --exclude named could be a cheap solution?)? I don't know :/.. 

I mean, we could try the above as a first step and see whether this takes care of things; I have a few good test cases we can try this on. Or we can just say defer this issue back to ODK, with @cmungall permission, and just use owltools for this particular thing until the resource situation is clarified. I dont mind either way!

**Comment by balhoff** [comment]
>  Does this mean, that maybe, relax should not relax equivalences between named classes

This makes sense to me. I don't see the value in it for the graph traversal use case. Of course when a named class is part of an expression (A and r some B), A would still get relaxed.

**Comment by cmungall** [comment]
I agree with Jim

**Comment by ignazio1977** [comment]
Hi, @matentzn pointed me at this issue to see if I had any clues as to what's going wrong.

I've played around with the code where the stack overflow issue happens, and I believe there's an issue in `RelatedObjectsHelper` - the gap spanning axioms are created from lists of relations, but when recurring to create the elements for these lists there is no test to check if the relation being added is already in the list. If it is, it means that particular path has already been explored and it's not necessary to continue the recursion.

I am guessing as to what the intent in the code is - haven't looked into it enough to be sure I'm not missing something important. However, if I'm right, the fix is straightforward and computationally inexpensive: simply stop the recursion once the relation being examined is already in the list of known relations. Replacing the list with a set slightly increases the memory requirements, but it also removes the creation of duplicate axioms that would happen even in a DAG scenario, and it allows the code to deal with cycles as the ones found here (@matentzn example finishes running in a few seconds on my very dated laptop, with default memory allocation).

The patch implementing my fix is as follows:

diff --git a/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java b/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java
index bc9a40a..ec4a80d 100644
--- a/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java
+++ b/robot-core/src/main/java/org/obolibrary/robot/RelatedObjectsHelper.java
@@ -770,10 +770,10 @@
    * @return set of OWLAxioms to maintain hierarchy
    */
   public static Set<OWLAxiom> spanGaps(OWLOntology ontology, Set<OWLObject> objects) {
-    List<Map<String, OWLAnnotationProperty>> aPropPairs = new ArrayList<>();
-    List<Map<String, OWLClassExpression>> classPairs = new ArrayList<>();
-    List<Map<String, OWLDataPropertyExpression>> dPropPairs = new ArrayList<>();
-    List<Map<String, OWLObjectPropertyExpression>> oPropPairs = new ArrayList<>();
+    Set<Map<String, OWLAnnotationProperty>> aPropPairs = new HashSet<>();
+    Set<Map<String, OWLClassExpression>> classPairs = new HashSet<>();
+    Set<Map<String, OWLDataPropertyExpression>> dPropPairs = new HashSet<>();
+    Set<Map<String, OWLObjectPropertyExpression>> oPropPairs = new HashSet<>();
     Set<OWLAxiom> axioms = new HashSet<>();
     OWLDataFactory dataFactory = OWLManager.getOWLDataFactory();
 
@@ -1342,7 +1342,7 @@
   private static void spanGapsHelper(
       OWLOntology ontology,
       Set<OWLObject> objects,
-      List<Map<String, OWLAnnotationProperty>> propPairs,
+      Set<Map<String, OWLAnnotationProperty>> propPairs,
       OWLAnnotationProperty property,
       Collection<OWLAnnotationProperty> superProperties) {
     for (OWLAnnotationProperty sp : superProperties) {
@@ -1351,14 +1351,15 @@
           Map<String, OWLAnnotationProperty> propertyPair = new HashMap<>();
           propertyPair.put(SUB, property);
           propertyPair.put(SUPER, sp);
-          propPairs.add(propertyPair);
-          property = sp;
-          spanGapsHelper(
-              ontology,
-              objects,
-              propPairs,
-              property,
-              EntitySearcher.getSuperProperties(property, ontology));
+          if (propPairs.add(propertyPair)) {
+            property = sp;
+            spanGapsHelper(
+                ontology,
+                objects,
+                propPairs,
+                property,
+                EntitySearcher.getSuperProperties(property, ontology));
+          }
         }
       } else if (!sp.isAnonymous()) {
         spanGapsHelper(
@@ -1385,22 +1386,22 @@
   private static void spanGapsHelper(
       OWLOntology ontology,
       Set<OWLObject> objects,
-      List<Map<String, OWLClassExpression>> classPairs,
-      OWLClass cls,
+      Set<Map<String, OWLClassExpression>> classPairs, OWLClass cls,
       Collection<OWLClassExpression> superClasses) {
     for (OWLClassExpression sc : superClasses) {
       if (objects.containsAll(sc.getSignature())) {
         Map<String, OWLClassExpression> classPair = new HashMap<>();
         classPair.put(SUB, cls);
         classPair.put(SUPER, sc);
-        classPairs.add(classPair);
-        if (!sc.isAnonymous()) {
-          spanGapsHelper(
-              ontology,
-              objects,
-              classPairs,
-              sc.asOWLClass(),
+        if (classPairs.add(classPair)) {
+          // only recurse if the pair just added was not already present in the set.
+          if (!sc.isAnonymous()) {
+            spanGapsHelper(ontology,
+                objects,
+                classPairs,
+                sc.asOWLClass(),
               getSuperClasses(ontology, sc.asOWLClass()));
+          }
         }
       } else if (!sc.isAnonymous()) {
         spanGapsHelper(
@@ -1423,7 +1424,7 @@
   private static void spanGapsHelper(
       OWLOntology ontology,
       Set<OWLObject> objects,
-      List<Map<String, OWLDataPropertyExpression>> propPairs,
+      Set<Map<String, OWLDataPropertyExpression>> propPairs,
       OWLDataProperty property,
       Collection<OWLDataPropertyExpression> superProperties) {
     for (OWLDataPropertyExpression sp : superProperties) {
@@ -1431,15 +1432,16 @@
         Map<String, OWLDataPropertyExpression> propertyPair = new HashMap<>();
         propertyPair.put(SUB, property);
         propertyPair.put(SUPER, sp);
-        propPairs.add(propertyPair);
-        if (!sp.isAnonymous()) {
-          property = (OWLDataProperty) sp;
-          spanGapsHelper(
-              ontology,
-              objects,
-              propPairs,
-              property,
-              EntitySearcher.getSuperProperties(property, ontology));
+        if (propPairs.add(propertyPair)) {
+          if (!sp.isAnonymous()) {
+            property = (OWLDataProperty) sp;
+            spanGapsHelper(
+                ontology,
+                objects,
+                propPairs,
+                property,
+                EntitySearcher.getSuperProperties(property, ontology));
+          }
         }
       } else if (!sp.isAnonymous()) {
         spanGapsHelper(
@@ -1466,7 +1468,7 @@
   private static void spanGapsHelper(
       OWLOntology ontology,
       Set<OWLObject> objects,
-      List<Map<String, OWLObjectPropertyExpression>> propPairs,
+      Set<Map<String, OWLObjectPropertyExpression>> propPairs,
       OWLObjectProperty property,
       Collection<OWLObjectPropertyExpression> superProperties) {
     for (OWLObjectPropertyExpression sp : superProperties) {
@@ -1474,15 +1476,16 @@
         Map<String, OWLObjectPropertyExpression> propertyPair = new HashMap<>();
         propertyPair.put(SUB, property);
         propertyPair.put(SUPER, sp);
-        propPairs.add(propertyPair);
-        if (!sp.isAnonymous()) {
-          property = (OWLObjectProperty) sp;
-          spanGapsHelper(
-              ontology,
-              objects,
-              propPairs,
-              property,
-              EntitySearcher.getSuperProperties(property, ontology));
+        if (propPairs.add(propertyPair)) {
+          if (!sp.isAnonymous()) {
+            property = (OWLObjectProperty) sp;
+            spanGapsHelper(
+                ontology,
+                objects,
+                propPairs,
+                property,
+                EntitySearcher.getSuperProperties(property, ontology));
+          }
         }
       } else if (!sp.isAnonymous()) {
         spanGapsHelper(

**Comment by rctauber** [comment]
Thanks @ignazio1977! Sorry for the delay in response, we've been getting ready for a conference. This patch works great at resolving the stack overflow. I'm not sure why I didn't use sets in the first place... silly oversight. It looks like all our tests still pass as well.

**Comment by beckyjackson** [comment]
This was resolved by #547 - please reopen if there are any other issues.
